kornia/geometry/liegroup/se2.py

```diff
@@ -0,0 +1,252 @@
+# kornia.geometry.se2 module inspired by Sophus-sympy.
+# https://github.com/strasdat/Sophus/blob/master/sympy/sophus/se2.py
+from typing import Optional, Tuple, Union
+
+from kornia.core import Module, Parameter, Tensor, concatenate, pad, rand, stack, tensor, where
+from kornia.geometry.liegroup._utils import check_se2_r_t_shape, check_v_shape
+from kornia.geometry.liegroup.so2 import So2
+from kornia.testing import KORNIA_CHECK, KORNIA_CHECK_TYPE
+
+
+class Se2(Module):
+    r"""Base class to represent the Se2 group.
+
+    The SE(2) is the group of rigid body transformations about the origin of two-dimensional Euclidean
+    space :math:`R^2` under the operation of composition.
+    See more:
+
+    Example:
+        >>> so2 = So2.identity(1)
+        >>> t = torch.ones((1, 2))
+        >>> se2 = Se2(so2, t)
+        >>> se2
+        rotation: Parameter containing:
+        tensor([1.+0.j], requires_grad=True)
+        translation: Parameter containing:
+        tensor([[1., 1.]], requires_grad=True)
+    """
+
+    def __init__(self, rotation: So2, translation: Tensor) -> None:
+        """Constructor for the base class.
+
+        Internally represented by a complex number `z` and a translation 2-vector.
+
+        Args:
+            r: So2 group encompassing a rotation.
+            t: translation vector with the shape of :math:`(B, 2)`.
+
+        Example:
+            >>> so2 = So2.identity(1)
+            >>> t = torch.ones((1, 2))
+            >>> se2 = Se2(so2, t)
+            >>> se2
+            rotation: Parameter containing:
+            tensor([1.+0.j], requires_grad=True)
+            translation: Parameter containing:
+            tensor([[1., 1.]], requires_grad=True)
+        """
+        super().__init__()
+        KORNIA_CHECK_TYPE(rotation, So2)
+        # TODO change to KORNIA_CHECK_SHAPE once there is multiple shape support
+        check_se2_r_t_shape(rotation, translation)
+        self._rotation = rotation
+        self._translation = Parameter(translation)
+
+    def __repr__(self) -> str:
+        return f"rotation: {self.r}\ntranslation: {self.t}"
+
+    def __getitem__(self, idx) -> 'Se2':
```

Comment:
here we should have an assert to check the idx based on the batch size (if any)

Reply:
something like this?
`KORNIA_CHECK(idx >= 0, msg="idx must be positive")`
if its unbatched it should give an out of index error

Reply:
more like:
```python
KORNIA_CHECK(len(self.rotation.shape) >= 3)
KORNIA_CHECK(idx >= 0 and idx < self.rotation.shape[0])
```

Reply:
Note that there's no rotation.shape but rotation.z.shape. `len(self.rotation.z.shape)` will return 1 if its batched, 0 if unbatched.  The current setup will produce the following errors for out of bound indexing:
```
se2 = Se2.random()  
se2
rotation: Parameter containing:
tensor(0.6863+0.0623j, requires_grad=True)
translation: Parameter containing:
tensor([0.1413, 0.9721], requires_grad=True)
se2[0]
*** IndexError: invalid index of a 0-dim tensor. Use `tensor.item()` in Python or `tensor.item<T>()` in C++ to convert a 0-dim tensor to a number
se2 = Se2.random(2)  
se2[2]
*** IndexError: index 2 is out of bounds for dimension 0 with size 2
se2[-1]
rotation: Parameter containing:
tensor([0.1879+0.3637j], requires_grad=True)
translation: Parameter containing:
tensor([[0.9053, 0.0592]], requires_grad=True)
se2[1]
rotation: Parameter containing:
tensor([0.1879+0.3637j], requires_grad=True)
translation: Parameter containing:
tensor([[0.9053, 0.0592]], requires_grad=True)

```
Moreover the negative indexing is congruent with np indexing as well

Reply:
Yeah, actually why translation is batched here?

Reply:
yea, it shouldn't. I'll take a look